engineering-decision-records/010-ledger-validation-context.md

```diff
@@ -0,0 +1,321 @@
+
+### 2. Interleaving of validations and I/O state changes
+
+The second challenge stems from using an on-disk solution to manage this state effectively. Given the tight time constraints that the validation demands, performing an I/O operation for every state change isn't realistic. Plus, the interleaving of validation and I/O operations with side effects complicates the error handling and makes the overall code hard to test and instrument.
+
+### 3. Additional context and extra information lookups
+
+Besides, the transition rules cannot solely run from the local context of a block or transaction. In most cases, additional information is required. For example, the output address and values corresponding to a spent input aren't part of the transaction but are present in the UTxO state.
+
+Furthermore, it isn't possible (or desirable) to load the entire ledger state in memory and operate from it. This defies the very design goals of the on-disk ledger state. In practice, it is also unnecessary. Even though transactions require an extra context, that context is negligible in front of the entire ledger state.
+
+### 4. Validations-as-a-library
+
+There's an explicit desire to keep the block/transaction validation rules as lean as possible and free of any storage or state management concerns. The use case here resides in providing better libraries for client applications, which could leverage battle-tested rules from full-node implementations to pre-validate transactions and provide a better user experience without re-implementing such validations.
+
+From an audit standpoint, keeping the validation rules lean enables a better understanding of the overall ledger's behaviour, which can otherwise be rapidly obfuscated in implementation details and complex data structure updates.
+
+## Decision
+
+### Abstract away state management from validations
+
+To decouple validations from state-management, we need to _at least_ abstract one or the other. After a few debates, our choice leaned towards abstracting the state management from the validations and keeping validation rules as the primary logic driver. We, therefore, provide an abstract interface as an eDSL available to the validation rules. The concrete implementation of those state operations is specified independently and chosen at the top of the stack by the caller.
+
+For example, we can capture the operation around UTxO management as follows:
+
+```rust
+pub trait UtxoStateManagement {
+    fn consume(&mut self, input: TransactionInput);
+    fn produce(&mut self, input: TransactionInput, output: TransactionOutput);
```

Comment:
`produce` doesn't convey the mutation imo. Maybe something like `append` or `include`?